File: app/src/main/java/com/stokey/designpattern/creator/singleton/Singleton.java
package com.stokey.designpattern.creator.singleton;

/**
 * Created by stokey on 2017/8/5.
 */

public class Singleton {

    /**
     * 饿汉单例模式
     */
    static class PlainOldSingleton {
        private static Singleton INSTANCE = new Singleton();

        private PlainOldSingleton() {
        }

        public static Singleton getInstance() {
            return INSTANCE;
        }

    }

    /**
     * 懒汉单例模式（变种）,非线程安全
     */
    static class LazyNotThreadSafe {
        private static LazyNotThreadSafe INSTANCE;

        private LazyNotThreadSafe() {
        }

        public static LazyNotThreadSafe getInstance() {
            if (INSTANCE == null) {
                INSTANCE = new LazyNotThreadSafe();
            }
            return INSTANCE;
        }
    }

    /**
     * 懒汉单例模式,线程安全
     */
    static class LazyThreadSafeSynchronized {
        private static LazyThreadSafeSynchronized INSTANCE;

        private LazyThreadSafeSynchronized() {
        }

        public static synchronized LazyThreadSafeSynchronized getInstance() {
            if (INSTANCE == null) {
                INSTANCE = new LazyThreadSafeSynchronized();
            }
            return INSTANCE;
        }
    }

    /**
     * 双重校验锁
     */
    static class LazyThreadSafeDoubleCheck {
        private static volatile LazyThreadSafeDoubleCheck INSTANCE;

        private LazyThreadSafeDoubleCheck() {
        }

        public static LazyThreadSafeDoubleCheck getInstance() {
            if (INSTANCE == null) {
                synchronized (LazyThreadSafeDoubleCheck.class) {
                    if (INSTANCE == null) {
                        INSTANCE = new LazyThreadSafeDoubleCheck();
                    }
                }
            }
            return INSTANCE;
        }
    }

    /**
     * 静态内部类
     */
    static class LazyThreadSafeStaticInnerClass {
        private static class Holder {
            private static LazyThreadSafeStaticInnerClass INSTANCE = new LazyThreadSafeStaticInnerClass();
        }

        private LazyThreadSafeStaticInnerClass() {
        }

        public static LazyThreadSafeStaticInnerClass getIntance() {
            return Holder.INSTANCE;
        }
    }

    /**
     * 枚举
     */
    static enum EnumSingleton {
        INSTANCE;

        public static EnumSingleton getInstance() {
            return INSTANCE;
        }
    }
}

File: app/src/main/java/com/stokey/designpattern/creator/singleton/Singleton.kt
package com.stokey.designpattern.creator.singleton

/**
 * Created by stokey on 2017/8/5.
 */

/**
 * 饿汉单例模式
 */
object PlainOldSingleton{

}

/**
 * 懒汉单例模式，非线程安全
 */
class LazyNotThreadSafe{
    companion object {
        val instance by lazy(LazyThreadSafetyMode.NONE) {
            LazyNotThreadSafe()
        }

        // Java等价写法，获取单利使用get方法
        private var instance2: LazyNotThreadSafe? = null

        fun get(): LazyNotThreadSafe {
            if (instance2 == null) {
                instance2 = LazyNotThreadSafe()
            }
            return instance2!!
        }
    }
}

/**
 * 懒汉单例模式，线程安全
 */
class LazyThreadSafeSynchronized{
    companion object {
        private var instance: LazyThreadSafeSynchronized? = null

        @Synchronized
        fun get(): LazyThreadSafeSynchronized {
            if (instance == null) instance = LazyThreadSafeSynchronized()
            return instance!!
        }
    }
}

/**
 * 双重校验
 */
class LazyThreadSafeDoubleCheck {
    companion object {
        val instance by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            LazyThreadSafeDoubleCheck()
        }

        // Java等价方法，使用get()方法调用单例
        private @Volatile var instance2: LazyThreadSafeDoubleCheck? = null
        fun get(): LazyThreadSafeDoubleCheck {
            if (instance2 == null){
                synchronized(this){
                    if (instance2 == null){
                        instance2 = LazyThreadSafeDoubleCheck()
                    }
                }
            }
            return instance2!!
        }

    }
}

/**
 * 静态内部类单例模式
 */
class LazyThreadSafeStaticInnerObject{
    private object Holder{
        val instance = LazyThreadSafeStaticInnerObject()
    }

    companion object {
        fun getInstance() = Holder.instance
    }
}

/**
 * 枚举单例模式
 */
enum class EnumSingleton{
    INSTANCE;
    fun get():EnumSingleton{
        return INSTANCE
    }
}